File: ufoagentd/controlplane.py
"""Thin stdlib HTTP client for the UFOAgent control plane API."""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any

from . import __version__


class ControlPlaneError(Exception):
    def __init__(self, status: int, code: str, message: str):
        super().__init__(f"{status} {code}: {message}")
        self.status = status
        self.code = code


@dataclass
class Credential:
    provider: str
    base_url: str
    model: str
    api_key: str
    expires_at: int
    lease_id: str

    @staticmethod
    def from_json(d: dict[str, Any]) -> "Credential":
        return Credential(
            provider=d["provider"], base_url=d["base_url"], model=d["model"],
            api_key=d["api_key"], expires_at=int(d["expires_at"]), lease_id=d["lease_id"],
        )


class ControlPlane:
    def __init__(self, base_url: str, token: str | None = None, timeout: float = 30.0):
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.timeout = timeout

    def _request(self, method: str, path: str, body: dict | None = None, auth: bool = False) -> dict[str, Any]:
        url = f"{self.base_url}{path}"
        data = json.dumps(body).encode("utf-8") if body is not None else None
        # Explicit User-Agent: the default "Python-urllib/x" is blocked by Cloudflare's bot rules.
        headers = {"accept": "application/json", "user-agent": f"ufoagentd/{__version__}"}
        if data is not None:
            headers["content-type"] = "application/json"
        if auth:
            if not self.token:
                raise ControlPlaneError(401, "no_token", "Agent is not linked (no token)")
            headers["authorization"] = f"Bearer {self.token}"
        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode("utf-8") or "{}")
        except urllib.error.HTTPError as e:
            payload: dict[str, Any] = {}
            try:
                payload = json.loads(e.read().decode("utf-8"))
            except Exception:
                pass
            raise ControlPlaneError(e.code, payload.get("error", "http_error"), payload.get("message", str(e)))
        except urllib.error.URLError as e:
            raise ControlPlaneError(0, "network_error", str(e.reason))

    # ---- agent endpoints ----
    def version(self) -> dict[str, Any]:
        return self._request("GET", "/v1/agent/version")

    def link_start(self, agent_name: str, platform: str) -> dict[str, Any]:
        return self._request("POST", "/v1/link/start", {"agent_name": agent_name, "platform": platform})

    def link_poll(self, device_code: str) -> dict[str, Any]:
        return self._request("POST", "/v1/link/poll", {"device_code": device_code})

    def get_credentials(self) -> Credential:
        return Credential.from_json(self._request("GET", "/v1/credentials", auth=True))

    def heartbeat(self, agent_version: str, platform: str) -> dict[str, Any]:
        return self._request("POST", "/v1/heartbeat", {"agent_version": agent_version, "platform": platform}, auth=True)

File: src/controlplane.rs
//! Thin HTTP client for the UFOAgent control plane API (ureq, sync).

use anyhow::{anyhow, bail, Result};
use serde::Deserialize;

/// Explicit User-Agent: the default agent string is blocked by Cloudflare's bot rules (403).
pub const USER_AGENT: &str = concat!("ufoagent/", env!("CARGO_PKG_VERSION"));

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct Credential {
    pub provider: String,
    pub base_url: String,
    pub model: String,
    pub api_key: String,
    pub expires_at: i64,
    pub lease_id: String,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct LinkStart {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    #[serde(default)]
    pub verification_uri_complete: Option<String>,
    #[serde(default)]
    pub interval: Option<u64>,
    #[serde(default)]
    pub expires_in: Option<u64>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct Heartbeat {
    #[serde(default)]
    pub ok: bool,
    #[serde(default)]
    pub min_version: Option<String>,
    #[serde(default)]
    pub server_time: Option<i64>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
pub struct LinkPoll {
    pub status: String,
    #[serde(default)]
    pub agent_id: Option<String>,
    #[serde(default)]
    pub agent_token: Option<String>,
    #[serde(default)]
    pub interval: Option<u64>,
}

pub struct ControlPlane {
    base: String,
    token: Option<String>,
}

impl ControlPlane {
    pub fn new(base: &str, token: Option<String>) -> Self {
        ControlPlane {
            base: base.trim_end_matches('/').to_string(),
            token,
        }
    }

    fn url(&self, path: &str) -> String {
        format!("{}{}", self.base, path)
    }

    fn get_json<T: for<'de> Deserialize<'de>>(&self, path: &str, auth: bool) -> Result<T> {
        let mut req = ureq::get(&self.url(path))
            .set("user-agent", USER_AGENT)
            .set("accept", "application/json");
        if auth {
            req = req.set("authorization", &format!("Bearer {}", self.bearer()?));
        }
        Self::send(req.call())
    }

    fn post_json<T: for<'de> Deserialize<'de>>(
        &self,
        path: &str,
        body: serde_json::Value,
        auth: bool,
    ) -> Result<T> {
        let mut req = ureq::post(&self.url(path))
            .set("user-agent", USER_AGENT)
            .set("accept", "application/json");
        if auth {
            req = req.set("authorization", &format!("Bearer {}", self.bearer()?));
        }
        Self::send(req.send_json(body))
    }

    fn bearer(&self) -> Result<&str> {
        self.token
            .as_deref()
            .ok_or_else(|| anyhow!("agent is not linked (no token)"))
    }

    fn send<T: for<'de> Deserialize<'de>>(
        resp: std::result::Result<ureq::Response, ureq::Error>,
    ) -> Result<T> {
        match resp {
            Ok(r) => Ok(r.into_json::<T>()?),
            Err(ureq::Error::Status(code, r)) => {
                let body = r.into_string().unwrap_or_default();
                bail!(
                    "control plane HTTP {}: {}",
                    code,
                    body.chars().take(300).collect::<String>()
                )
            }
            Err(e) => bail!("control plane request failed: {}", e),
        }
    }

    pub fn version(&self) -> Result<serde_json::Value> {
        self.get_json("/v1/agent/version", false)
    }

    pub fn link_start(&self, agent_name: &str, platform: &str) -> Result<LinkStart> {
        self.post_json(
            "/v1/link/start",
            serde_json::json!({ "agent_name": agent_name, "platform": platform }),
            false,
        )
    }

    pub fn link_poll(&self, device_code: &str) -> Result<LinkPoll> {
        self.post_json(
            "/v1/link/poll",
            serde_json::json!({ "device_code": device_code }),
            false,
        )
    }

    pub fn get_credentials(&self) -> Result<Credential> {
        self.get_json("/v1/credentials", true)
    }

    pub fn heartbeat(&self, agent_version: &str, platform: &str) -> Result<Heartbeat> {
        self.post_json(
            "/v1/heartbeat",
            serde_json::json!({ "agent_version": agent_version, "platform": platform }),
            true,
        )
    }
}
